odl/solvers/nonsmooth/adupdates.py

```diff
@@ -0,0 +1,177 @@
+
+    * Assume that `f_j` is a `SeparableSum`, then setting `majs[i]` a list
+    :math:`(\gamma_1, \ldots, \gamma_k)`of positive floats corresponds to the
+    choice of a block-diagonal matrix :math:`M_j`, where each block corresponds
+    to one of the space components and equals :math:`\gamma_i \mathrm{Id}`.
+
+    * Assume that `f_j` is a `L1_norm` or a `L2_norm_squared`, then setting
+    `majs[i]` a `f_j.domain.element` :math:`z` corresponds to the choice
+    :math:`\mathrm{diag}(z)`.
+
+    References
+    ----------
+    [MF2015] McGaffin, M G, and Fessler, J A. *Alternating dual updates
+    algorithm for X-ray CT reconstruction on the GPU*. IEEE Transactions
+    on Computational Imaging, 1.3 (2015), pp 186--199.
+    """
+    # Check the lenghts of the lists (= number of dual variables)
+    length = len(funcs)
+    if not len(funcs) == len(ops):
+        raise ValueError('Number of `ops` {} and number '
+                         'of `funcs` {} do not agree.'
+                         ''.format(len(ops), len(funcs)))
+
+    # Check if operators have a common domain
+    # (the space of the primal variable):
+    domain = ops[0].domain
+    if any(domain != opi.domain for opi in ops):
+        raise ValueError('Domains of `ops` are not all equal.')
+
+    # Check if range of the operators equals domain of the functionals
+    ranges = [opi.range for opi in ops]
+    if any(ops[i].range != funcs[i].domain for i in range(length)):
+        raise ValueError('Ranges of operators and domains'
+                         'of functionals do not agree.')
+
+    # Initialization of the dual variables
+    duals = [space.zero() for space in ranges]
+
+    # Reusable elements in the ranges, one per type of space
+    unique_ranges = set(ranges)
+    tmp_rans = {ran: ran.element() for ran in unique_ranges}
+
+    # Prepare the proximal operators. Since the stepsize does not vary over
+    # the iterations, we always use the same proximal operator.
+    proxs = [func.convex_conj.proximal(stepsize / maj)
+             for (func, maj) in zip(funcs, majs)]
+
+    # Iteratively find a solution
+    for _ in range(niter):
+        # Update x = x - 1/stepsize * sum([ops[i].adjoint(duals[i])
+        # for i in range(length)])
+        for i in range(length):
+            x -= 1.0 / stepsize * ops[i].adjoint(duals[i])
+
+        if random:
+            rng = np.random.permutation(range(len(ops)))
+        else:
+            rng = range(len(ops))
+
+        for i in rng:
+            step = stepsize / majs[i]
+            arg = duals[i] + step * ops[i](x)
+            tmp_ran = tmp_rans[ops[i].range]
+            proxs[i](arg, out=tmp_ran)
+            x -= 1.0 / stepsize * ops[i].adjoint(tmp_ran - duals[i])
+            duals[i] = tmp_ran.copy()
+
+            if callback is not None and callback_loop == 'inner':
+                callback(x)
+        if callback is not None and callback_loop == 'outer':
+            callback(x)
```

Comment:
This is a pretty complex optimization method. I'm sure it's a good one.
What I can recommend is to make a "dead simple" implementation of the same optimizer that does not do any fancy stuff like caching, in-place operation of optimized arithmetic. Just with as little code as possible. That's really good for debugging. See for example [here](https://github.com/odlgroup/odl/blob/87e04941f4337de190654897691327f8964c0f03/odl/solvers/nonsmooth/admm.py#L149-L163).